src/project.cpp

```diff
@@ -1907,46 +1930,68 @@ Project::DataQualifiers Project::getDataQualifiers(QString text, QString label)
+void Project::insertTilesetLabel(QString label, QString isSecondaryStr) {
+    bool ok;
+    bool isSecondary = ParseUtil::gameStringToBool(isSecondaryStr, &ok);
+    if (!ok) {
+        logError(QString("Unable to convert value '%1' of isSecondary to bool for tileset %2.").arg(isSecondaryStr).arg(label));
+        return;
     }
+    insertTilesetLabel(label, isSecondary);
+}
 
-    QRegularExpression re("(?<label>[A-Za-z0-9_]*):{1,2}[A-Za-z0-9_@ ]*\\s+.+\\s+\\.byte\\s+(?<isSecondary>[A-Za-z0-9_]+)");
-    QRegularExpressionMatchIterator iter = re.globalMatch(headers_text);
-
-    while (iter.hasNext()) {
-        QRegularExpressionMatch match = iter.next();
-        QString tilesetLabel = match.captured("label");
-        QString secondaryTilesetValue = match.captured("isSecondary");
+bool Project::readTilesetLabels() {
+    QStringList primaryTilesets;
+    QStringList secondaryTilesets;
+    this->tilesetLabels.clear();
+    this->tilesetLabels.insert("primary", primaryTilesets);
+    this->tilesetLabels.insert("secondary", secondaryTilesets);
+    this->tilesetLabelsOrdered.clear();
 
-        if (secondaryTilesetValue != "1" && secondaryTilesetValue != "TRUE" 
-         && secondaryTilesetValue != "0" && secondaryTilesetValue != "FALSE") {
-            logWarn(QString("Unexpected secondary tileset flag found in %1. Expected 'TRUE', 'FALSE', '1', or '0', but found '%2'")
-                    .arg(tilesetLabel).arg(secondaryTilesetValue));
-            continue;
+    QString filename = projectConfig.getFilePath(ProjectFilePath::tilesets_headers);
+    QFileInfo fileInfo(this->root + "/" + filename);
+    if (!fileInfo.exists() || !fileInfo.isFile()) {
+        // If the tileset headers file is missing, the user may still have the old assembly format.
+        this->usingAsmTilesets = true;
+        QString asm_filename = projectConfig.getFilePath(ProjectFilePath::tilesets_headers_asm);
+        QString text = parser.readTextFile(this->root + "/" + asm_filename);
+        if (text.isEmpty()) {
+            logError(QString("Failed to read tileset labels from '%1' or '%2'.").arg(filename).arg(asm_filename));
+            return false;
+        }
+        QRegularExpression re("(?<label>[A-Za-z0-9_]*):{1,2}[A-Za-z0-9_@ ]*\\s+.+\\s+\\.byte\\s+(?<isSecondary>[A-Za-z0-9_]+)");
+        QRegularExpressionMatchIterator iter = re.globalMatch(text);
+        while (iter.hasNext()) {
+            QRegularExpressionMatch match = iter.next();
+            insertTilesetLabel(match.captured("label"), match.captured("isSecondary"));
+        }
+        filename = asm_filename; // For error reporting further down
+    } else {
+        this->usingAsmTilesets = false;
+        const auto structs = parser.readCStructs(filename, "", Tileset::getHeaderMemberMap(this->usingAsmTilesets));
+        QStringList labels = structs.keys();
+        // TODO: This is alphabetical, AdvanceMap import wants the vanilla order in tilesetLabelsOrdered
```

Comment:
Just checking, did you actually test the AdvanceMap import functionality with these changes, or is this hypothetical.  I haven't tested that in a long time, so I'm wondering if it still even works.

Reply:
Hypothetical, based on my interpretation of this code
<https://github.com/huderlem/porymap/blob/2d87a6222a8cba1c5d5439a578235e4d04f6dfeb/src/core/mapparser.cpp#L78-L88>

Reply:
Ok, I'll test it just for sanity

Reply:
This is still working.  The user gets to choose the tileset in the import dialog anyway, so it's not a big deal.